### PR TITLE
move Rust crates to their own subdirectories

### PR DESCRIPTION
**Pull request by Be-ing**

### CODE REVIEW

**Review by ahayzen-kdab — CHANGES_REQUESTED** [review]
Looks good, but i think `book/src/concepts/qqmlextensionplugin.md` needs updating ? As it mentions a `core` folder that has now been removed ?

**Comment by LeonMatthes** [comment]
Hm, I'm uncertain this is something we actually want.

I rather like how all rs, qml and c++ files co-exist in the same directories.
It really emphasises the polyglot nature of our cxx-qt

**Comment by Be-ing** [comment]
When I first started working on cxx-qt, I found the present organization of the files confusing because it's not immediately obvious what parts CMake is handling versus what parts Cargo is handling. Also, in real world, nontrivial applications, there's probably going to be much more C++ code than Rust, so this directory structure models what would be more appropriate in that context.

**Comment by LeonMatthesKDAB** [comment]
Hm, its a fair point that larger applications might favor a `rust` directory for all the Rust code.

However, our repository only contains small codebases. And for those I think it makes sense to keep it as-is.
Otherwise we almost end up with more subfolders than actual files :see_no_evil: 

We could stress in the book though that there are different ways to organize a repository and that our examples are just that: minimal examples.

**Comment by Be-ing** [comment]
I think our examples should resemble how we recommend to use cxx-qt in real applications as much as practical. Otherwise we would have `find_package(Corrosion)` and `find_package(QT)` repeated in every examples' CMakeLists.txt.

**Comment by ahayzen-kdab** [comment]
> Hm, its a fair point that larger applications might favor a `rust` directory for all the Rust code.

Or we have a split of some examples with the directory and some not, eg the qml_minimal should really be ... minimal ? So not have the sub directories ? Then before the qml_extension_plugin already had some splitting of folders and would be closer to a larger real world application. :thinking:

Do we need this split in the tests/ folders?

I also should reevaluate the examples/ and tests/ as there are some duplication and there is some testing occurring in the examples :') And some of them could probably be merged like the threaded_logic examples into qml_features then have a threading.rs example etc. But something for later ...

**Comment by Be-ing** [comment]
I think it would get confusing for both us and people looking at the examples as guides for their projects if we had different directory organization for different examples.

**Comment by LeonMatthesKDAB** [comment]
I can see the argument to go for a recommendation of how to lay out a CXX-Qt repository.
In that case I think we should hoist the `rust` and tests directory up one level, i.e.:
- qml_minimal
  - CMakeLists.txt
  - rust/
     - *Rust sources*
  - src/ - maybe call this `cpp` or `qt`
    - *Qt sources*
  - tests/
    - *Tests*
At the moment the "main level" is rather empty with just the CMakeLists.txt in it.

**Comment by ahayzen** [comment]
> I can see the argument to go for a recommendation of how to lay out a CXX-Qt repository. In that case I think we should hoist the `rust` and tests directory up one level, i.e.:
> 
> * qml_minimal
>   
>   * CMakeLists.txt
>   * rust/
>     
>     * _Rust sources_
>   * src/ - maybe call this `cpp` or `qt`
>     
>     * _Qt sources_
>   * tests/
>     
>     * _Tests_
> 
> At the moment the "main level" is rather empty with just the CMakeLists.txt and src folder in it.

I like that idea, could even then have
* rust/
* cpp/
* qml/ <- qrc and qml files
* tests/

Or
* rust/
* src/
* ui/
* tests/
🤔

**Comment by LeonMatthesKDAB** [comment]
> I like that idea, could even then have
> 
>     * rust/
>     * cpp/
>     * qml/ <- qrc and qml files
>     * tests/
> 
> 
> Or
> 
>     * rust/
>     * src/
>     * ui/
>     * tests/

In that case I'd go with the first option, to cleanly separate by language.

The differentiation between `src`, `ui` and `rust` is a bit unclear, as Rust is also part of your general "source", it just happens to be written in Rust. Even though I like the `ui` name.
So I'd go with either:
- rust/
- cpp/
- ui/
- tests/
or:
- rust/
- cpp/
- qml/
- tests/

If you choose either of those, I can approve.

**Comment by Be-ing** [comment]
I like this:
```
cxx-qt/examples/demo_threading on  decouple_rust [✘!?⇕] via △ v3.22.2 via 🐍 v3.10.6 
❯ tree
.
├── bad_client.py
├── client.py
├── CMakeLists.txt
├── cpp
│   └── main.cpp
├── qml
│   ├── main.qml
│   └── qml.qrc
└── rust
    ├── build.rs
    ├── Cargo.toml
    └── src
        └── lib.rs
```

**Comment by Be-ing** [comment]
```
cxx-qt on  decouple_rust via △ v3.22.2 via 🦀 v1.63.0 
❯ tree examples tests
examples
├── cargo_without_cmake
│   ├── build.rs
│   ├── Cargo.toml
│   ├── CMakeLists.txt
│   ├── qml -> ../qml_minimal/qml/
│   └── src
│       ├── cpp
│       │   └── run.cpp
│       ├── cxxqt_object.rs -> ../../qml_minimal/rust/src/cxxqt_object.rs
│       └── main.rs
├── CMakeLists.txt
├── demo_threading
│   ├── bad_client.py
│   ├── client.py
│   ├── CMakeLists.txt
│   ├── cpp
│   │   └── main.cpp
│   ├── qml
│   │   ├── main.qml
│   │   └── qml.qrc
│   └── rust
│       ├── build.rs
│       ├── Cargo.toml
│       └── src
│           └── lib.rs
├── qml_extension_plugin
│   ├── CMakeLists.txt
│   ├── cpp
│   │   └── main.cpp
│   ├── plugin
│   │   ├── CMakeLists.txt
│   │   ├── cpp
│   │   │   └── plugin.cpp
│   │   ├── qmldir
│   │   ├── qmldir.license
│   │   └── rust
│   │       ├── build.rs
│   │       ├── Cargo.toml
│   │       └── src
│   │           └── lib.rs
│   ├── qml
│   │   ├── main.qml
│   │   └── qml.qrc
│   └── tests
│       ├── CMakeLists.txt
│       └── tst_myobject.qml
├── qml_features
│   ├── CMakeLists.txt
│   ├── cpp
│   │   └── main.cpp
│   ├── qml
│   │   ├── main.qml
│   │   └── qml.qrc
│   ├── rust
│   │   ├── build.rs
│   │   ├── Cargo.toml
│   │   └── src
│   │       ├── custom_base.rs
│   │       ├── empty.rs
│   │       ├── lib.rs
│   │       ├── mock_qt_types.rs
│   │       ├── rust_obj_invokables.rs
│   │       ├── serialisation.rs
│   │       ├── signals.rs
│   │       ├── struct_properties.rs
│   │       └── types.rs
│   └── tests
│       ├── myobject
│       │   ├── tst_myobject.cpp
│       │   └── tst_myobject.qml
│       ├── qttypes
│       │   ├── tst_qttypes.cpp
│       │   └── tst_qttypes.qml
│       └── serialisation
│           ├── tst_serialisation.cpp
│           └── tst_serialisation.qml
├── qml_minimal
│   ├── CMakeLists.txt
│   ├── cpp
│   │   └── main.cpp
│   ├── qml
│   │   ├── main.qml
│   │   └── qml.qrc
│   ├── rust
│   │   ├── build.rs
│   │   ├── Cargo.toml
│   │   └── src
│   │       ├── cxxqt_object.rs
│   │       └── lib.rs
│   └── tests
│       └── myobject
│           ├── tst_myobject.cpp
│           └── tst_myobject.qml
├── qml_with_threaded_logic
│   ├── CMakeLists.txt
│   ├── cpp
│   │   └── main.cpp
│   ├── qml
│   │   ├── main.qml
│   │   └── qml.qrc
│   ├── rust
│   │   ├── build.rs
│   │   ├── Cargo.toml
│   │   └── src
│   │       └── lib.rs
│   └── tests
│       └── website
│           ├── tst_website.cpp
│           └── tst_website.qml
└── README.md
tests
├── basic_cxx_only
│   ├── CMakeLists.txt
│   ├── cpp
│   │   ├── main.cpp
│   │   └── main.h
│   └── rust
│       ├── build.rs
│       ├── Cargo.toml
│       └── src
│           └── lib.rs
├── basic_cxx_qt
│   ├── CMakeLists.txt
│   ├── cpp
│   │   └── main.cpp
│   └── rust
│       ├── build.rs
│       ├── Cargo.toml
│       └── src
│           ├── data.rs
│           ├── lib.rs
│           └── types.rs
├── CMakeLists.txt
└── qt_types_standalone
    ├── CMakeLists.txt
    ├── cpp
    │   ├── bridge.h
    │   └── main.cpp
    └── rust
        ├── build.rs
        ├── Cargo.toml
        └── src
            └── lib.rs
```

File: book/src/concepts/nested_objects.md [hunk 1]
```diff
@@ -24,7 +24,7 @@ For use as a parameter in an invokable write `secondary_object: &mut crate::mymo
 The following example shows a nested object as a property and parameter.
 
 ```rust,ignore,noplayground
-{{#include ../../../examples/qml_features/src/nested.rs:book_macro_code}}
+#include ../../../examples/qml_features/rust/src/nested.rs:book_macro_code
```
Comment (Be-ing) [inline book/src/concepts/nested_objects.md:27]: I removed the `{{`/`}}` to get `mdbook build` to run without errors.

Comment (ahayzen-kdab) [reply book/src/concepts/nested_objects.md:27]: Interesting, that was in a commented block before as nested objects have been removed for now. But this is OK for now ..

**Review by ahayzen-kdab — APPROVED** [review]
LGTM :-)

**Pull request merged**